llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -4677,7 +4683,11 @@ VectorizationFactor LoopVectorizationPlanner::selectEpilogueVectorizationFactor(
     return Result;
   }
 
-  if (!CM.isEpilogueVectorizationProfitable(MainLoopVF)) {
+  unsigned Multiplier = IC;
+  if (MainLoopVF.isScalable())
+    Multiplier = getVScaleForTuning(OrigLoop, TTI).value_or(1);
```

Comment:
Possibly, but I didn't really look at SVE and since the remaining part of this change depends on using `RemainingIterations` for checking profitability, which has an (existing) `// TODO: extend to support scalable VFs.` I'd rather keep changing behavior for SVE as a follow-up.

Reply:
I think this change looks incorrect. Previously in `isEpilogueVectorizationProfitable` we did:

```
  unsigned Multiplier = 1;
  if (VF.isScalable())
    Multiplier = getVScaleForTuning(TheLoop, TTI).value_or(1);
  if ((Multiplier * VF.getKnownMinValue()) >= EpilogueVectorizationMinVF)
    return true;
```

i.e. for fixed-width VFs `Multiplier = 1`, whereas after this change `Multiplier = IC`. This is either biasing against or in favour of fixed-width VFs, which doesn't seem right. I think in order to match the previous behaviour the code should be:

```
  unsigned Multiplier = 1;
  if (MainLoopVF.isScalable())
    Multiplier = getVScaleForTuning(OrigLoop, TTI).
```

Reply:
Hi @fhahn can you verify this as well? I think it should be one of the following:

```
  unsigned Multiplier = IC;
  if (MainLoopVF.isScalable())
    Multiplier *= getVScaleForTuning(OrigLoop, TTI).value_or(1);
```

or 

```
  unsigned Multiplier = 1;
  if (MainLoopVF.isScalable())
    Multiplier = getVScaleForTuning(OrigLoop, TTI).value_or(1);
```

Reply:
I wouldn't say it is incorrect, it just keeps the original behavior for scalable vectors as I think @juliannagele doesn't have access to HW with scalable vectors, which would be needed to evaluate the impact of changing this for scalable vectors.

At this point already picked the VF for the main loop, so the only change is that we consider epilogue vectorization for more cases with fixed vectors.

To avoid regression with fixed vectors, this patch relies on code that checks the number of remaining iterations, which currently doesn't support scalable vectors (look for `// TODO: extend to support scalable VFs`), which probably should be fixed first. Again, this should probably be done by someone with access to HW supporting scalable vectors.

Reply:
I can volunteer for that, but I am really keen that this lands first. :)

Reply:
I believe this has now landed, and #100385 is hopefully unblocked. Dave put up #116607 for changing this part too.

Reply:
Ah, okay, didn't realise that, thanks!